pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -57,50 +58,80 @@ cdef inline object create_timestamp_from_ts(int64_t value,
+    PLUS_INFTY = 1
+    NEAREST_HALF_EVEN = 2
+    NEAREST_HALF_PLUS_INFTY = 3
+    NEAREST_HALF_MINUS_INFTY = 4
+
+
+cdef inline _npdivmod(x1, x2):
+    """implement divmod for numpy < 1.13"""
+    return np.floor_divide(x1, x2), np.remainder(x1, x2)
+
+
+try:
+    from numpy import divmod as npdivmod
+except ImportError:
+    npdivmod = _npdivmod
+
+
+cdef inline _floor_int64(v, u):
+    return v - np.remainder(v, u)
+
+cdef inline _ceil_int64(v, u):
+    return v + np.remainder(-v, u)
+
+cdef inline _rounddown_int64(v, u):
+    return _ceil_int64(v - u//2, u)
+
+cdef inline _roundup_int64(v, u):
+    return _floor_int64(v + u//2, u)
+
+
+def round_nsint64(values, mode: RoundTo, freq):
     """
-    Applies rounding function at given frequency
+    Applies rounding mode at given frequency
 
     Parameters
     ----------
     values : :obj:`ndarray`
-    rounder : function, eg. 'ceil', 'floor', 'round'
+    mode : instance of `RoundTo` enumeration
     freq : str, obj
 
     Returns
     -------
     :obj:`ndarray`
     """
+
+    if not isinstance(mode, RoundTo):
+        raise ValueError('mode should be a RoundTo member')
+
     unit = to_offset(freq).nanos
 
-    # GH21262 If the Timestamp is multiple of the freq str
-    # don't apply any rounding
-    mask = values % unit == 0
-    if mask.all():
-        return values
-    r = values.copy()
-
-    if unit < 1000:
-        # for nano rounding, work with the last 6 digits separately
-        # due to float precision
-        buff = 1000000
-        r[~mask] = (buff * (values[~mask] // buff) +
-                    unit * (rounder((values[~mask] % buff) *
-                            (1 / float(unit)))).astype('i8'))
-    else:
-        if unit % 1000 != 0:
-            msg = 'Precision will be lost using frequency: {}'
-            warnings.warn(msg.format(freq))
-        # GH19206
-        # to deal with round-off when unit is large
-        if unit >= 1e9:
-            divisor = 10 ** int(np.log10(unit / 1e7))
-        else:
-            divisor = 10
-        r[~mask] = (unit * rounder((values[~mask] *
-                    (divisor / float(unit))) / divisor)
-                    .astype('i8'))
-    return r
+    if mode is RoundTo.MINUS_INFTY:
+        return _floor_int64(values, unit)
+    elif mode is RoundTo.PLUS_INFTY:
+        return _ceil_int64(values, unit)
+    elif mode is RoundTo.NEAREST_HALF_MINUS_INFTY:
+        return _rounddown_int64(values, unit)
+    elif mode is RoundTo.NEAREST_HALF_PLUS_INFTY:
+        return _roundup_int64(values, unit)
+    elif mode is RoundTo.NEAREST_HALF_EVEN:
+        # for odd unit there is no need of a tie break
+        if unit % 2:
+            return _rounddown_int64(values, unit)
+        d, r = npdivmod(values, unit)
+        mask = np.logical_or(
+            r > (unit // 2),
+            np.logical_and(r == (unit // 2), d % 2)
+        )
+        d[mask] += 1
+        return d * unit
+
+    raise NotImplementedError(mode)
```

Comment:
This code is unreachable and the `NotImplementedError` was cruft from a previous implementation.

In 44902695f4adba3406fb4750e819a5776683cae7 I inserted an `assert False` to avoid the code returning `None` if a bug is introduced by changing the `RoundTo` class.